_posts/2023-07-31-turbocharging-rust-code-verification.md

```diff
@@ -0,0 +1,405 @@
+
+For unions, just like for structs, the transform introduces a distinct variable for each field of the union. However, contrary to structs, the different fields of a union overlap in the byte-level layout. As a result, every time a union field gets assigned with a new value, all fields of the union are actually impacted, and the impact depends on how the fields overlap in the layout. This means that the variables representing the different fields of the union have to be handled as a group and globally updated after each update to any one of them.
+
+For instance, for a union defined as follows (using C syntax for simplicity):
+
+```c
+union {
+  unsigned long a;
+  unsigned int b;
+} u;
+```
+
+The byte-level layout of the fields is such that the lowest 4 bytes of `u.a` and all bytes of `u.b` overlap. As a result, updating `u.a` updates all bytes `u.b`, and updating `u.b` updates the lowest 4 bytes of `u.a`:
+
+```
+u       uuuuuuuuuuuuuuuu
+u.a     aaaaaaaaaaaaaaaa
+u.b             bbbbbbbb
+idx.   16       7      0
+      MSB             LSB
+```
+
+The transform introduces a variable `u_a` to represent `u.a`, and a variable `u_b` to represent `u.b`. These two variables are not independent, and every-time one of the fields is updated in the original program, both variables are updated in the transformed program.
+
+Applying the transform to the following program:
+
+```c
+int main() {
+  union {
+    unsigned long a;
+    unsigned int b;
+  } u;
+
+  u.a = 0x0000000000000000;
+  u.b = 0x87654321;
+  assert(u.a == 0x0000000087654321);
+  assert(u.b == 0x87654321);
+  return 0;
+}
+```
+
+produces the following transformed program:
+
+```c
+int main() {
+  unsigned long u_a;
+  unsigned int u_b;
+
+  // the bytes of u_b are equal the low bytes of u_a
+  u_b = (unsigned int) u_a;
+
+  // u.a = 0x0000000000000000;
+  u_a = 0x0000000000000000;
+  u_b = (unsigned int) 0x0000000000000000;
+
+  // u.b = 0x87654321;
+  u_a = (u_a & 0xFFFFFFFF00000000) | ((unsigned long) 0x87654321);
+  u_b = 0x87654321;
+
+  // assert(u.a == 0x0000000087654321);
+  assert(u_a == 0x0000000087654321);
+
+  // assert(u.b == 0x87654321);
+  assert(u_b == 0x87654321);
+  return 0;
+}
+```
+
+## Conclusion
+
+In conclusion, we would like to point out the multi-faceted approach to optimizing a tool such as Kani and making verification scalable in general. Some optimizations are geared towards solving unsolvable cases whereas others are more general in nature. A reflection of the difficult problem space we are dealing with. Specific tools may have the luxury of niche optimizations, but tools like Kani cannot afford that luxury. Kani has to be optimized at various levels to achieve realistic performant verification. We hope you enjoyed the blog post and get a sense of the need for a multi-faceted approach.
```

Comment:
Really liked how this conclusion invites to reflection, except for the sentence "Specific tools may have the luxury of niche optimizations, but tools like Kani cannot afford that luxury." Maybe I would change it just a little to say that "Kani has to be optimized at various levels" but also that "it requires us to innovate".

Reply:
reworded